scale_build/bootstrap/bootstrapdir.py

```diff
@@ -0,0 +1,149 @@
+            ]
+        )
+        self.setup_mounts()
+
+        if self.extra_packages_to_install:
+            self.run(['chroot', self.chroot_basedir, 'apt', 'install', '-y'] + self.extra_packages_to_install)
+
+        installed_packages = self.get_packages()
+
+        self.after_extra_packages_installation_steps()
+
+        # Save the correct repo in sources.list
+        apt_path = os.path.join(self.chroot_basedir, 'etc/apt')
+        apt_sources_path = os.path.join(apt_path, 'sources.list')
+        apt_sources = [f'deb {apt_repos["url"]} {apt_repos["distribution"]} {apt_repos["components"]}']
+
+        # Set bullseye repo as the priority
+        with open(os.path.join(apt_path, 'preferences'), 'w') as f:
+            f.write(get_apt_preferences())
+
+        # Add additional repos
+        for repo in apt_repos['additional']:
+            self.logger.debug('Adding additional repo: %r', repo['url'])
+            shutil.copy(os.path.join(BUILDER_DIR, repo['key']), os.path.join(self.chroot_basedir, 'apt.key'))
+            self.run(['chroot', self.chroot_basedir, 'apt-key', 'add', '/apt.key'])
+            os.unlink(os.path.join(self.chroot_basedir, 'apt.key'))
+            apt_sources.append(f'deb {repo["url"]} {repo["distribution"]} {repo["component"]}')
+
+        with open(apt_sources_path, 'w') as f:
+            f.write('\n'.join(apt_sources))
+
+        # Update apt
+        self.run(['chroot', self.chroot_basedir, 'apt', 'update'])
+
+        # Put our local package up at the top of the food chain
+        apt_sources.insert(0, 'deb [trusted=yes] file:/packages /')
+        with open(apt_sources_path, 'w') as f:
+            f.write('\n'.join(apt_sources))
+
+        self.clean_mounts()
+        self.save_build_cache(installed_packages)
+
+    def after_extra_packages_installation_steps(self):
+        pass
+
+    def add_trusted_apt_key(self):
+        self.run([
+            'apt-key', '--keyring', '/etc/apt/trusted.gpg.d/debian-archive-truenas-automatic.gpg', 'add',
+            os.path.join(BUILDER_DIR, 'keys/truenas.gpg')
+        ])
+
+    def run(self, *args, **kwargs):
+        return run(*args, logger=self.logger, **kwargs)
+
+    @property
+    def extra_packages_to_install(self):
+        raise NotImplementedError
+
+    @property
+    def deopts(self):
+        raise NotImplementedError
+
+    def setup_mounts(self):
+        self.run(['mount', 'proc', os.path.join(self.chroot_basedir, 'proc'), '-t', 'proc'])
+        self.run(['mount', 'sysfs', os.path.join(self.chroot_basedir, 'sys'), '-t', 'sysfs'])
+
+    def clean_mounts(self):
+        for command in (
+            ['umount', '-f', os.path.join(self.chroot_basedir, 'proc')],
+            ['umount', '-f', os.path.join(self.chroot_basedir, 'sys')],
+        ):
+            run(command, check=False)
+
+    def clean_setup(self):
+        self.clean_mounts()
+        shutil.rmtree(self.chroot_basedir, ignore_errors=True)
```

Comment:
What errors are you ignoring?

We expect `self.chroot_basedir` to be non-existent after this method terminates successfully. Will it always be so?

Reply:
I am not sure I follow, why it wouldn't be after this ? We ignore it so that we can safely call it before setting up bootstrap directory to make sure we have a clean start.

Reply:
If `shutil.rmtree` fails to remove something inside `self.chroot_basedir`, it won't be able to remove `self.chroot_basedir` too due to it being non-empty, but as `ignore_errors=True` is specified, this will be ignored and no exception will be raised.

Reply:
Understood. My only use case was to be able to call these endpoints even if the directories did not exist but yes, ignoring all errors is not wise. I'll make required changes for all such usages. Thank you for pointing it out.